docs/architecture/adr-033-protobuf-ocaps.md

```diff
@@ -0,0 +1,287 @@
+# ADR 033: Protobuf-based Inter-Module Object Capabilities
+
+## Changelog
+
+- 2020-10-05: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR introduces a system for permissioned inter-module communication or object capabilites (Ocaps) leveraging the
+protobuf `Query` and `Msg` service definitions defined in [ADR 021](./adr-021-protobuf-query-encoding.md) and
+[ADR 031](./adr-031-msg-service.md).
+
+## Context
+
+In the current Cosmos SDK documentation on the [Object-Capability Model](../docs/core/ocap.md), it is state that:
+
+> We assume that a thriving ecosystem of Cosmos-SDK modules that are easy to compose into a blockchain application will contain faulty or malicious modules.
+
+There is currently not a thriving ecosystem of Cosmos SDK modules. We hypothesize that this is in part due to:
+1. lack of a stable v1.0 Cosmos SDK to build modules off of. Module interfaces are changing, sometimes dramatically, from
+point release to point release, often for good reasons, but this does not create a stable foundation to build on.
+2. lack of a properly implemented object capability or even object-oriented encapsulation system.
```

Comment:
what about the capability module?

Reply:
I think it would be helpful to add what's the benefit of this model over [ADR 3](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-003-dynamic-capability-store.md). Also how would it replace it in the context of IBC (I assume we wouldn't like to have to ocap implementations).

Reply:
Good point. I can reference that in the ADR. But it isn't implemented universally across the SDK and I'm not sure how we would leverage it alone to deal with the lack of encapsulation in `x/bank` or `x/auth`.

Reply:
yeah, it was purposed for IBC, and the idea was to add support for other modules as well in the future. I think I saw a WIP implementation for the bank module using that. I think it would be useful if @AdityaSripal could review this ADR as he probably has the most context of ADR3 and x/capability .

Reply:
While this statement is probably true, I think it's a bit of stretch to state this as a reason why there is a lack of thriving modules. From my experience, SDK/module devs rarely care about this and just want to get their idea/app implemented as cleanly and quickly as possible. Most ppl don't care or even know about ocaps.

Reply:
> While this statement is probably true, I think it's a bit of stretch to state this as a reason why there is a lack of thriving modules. From my experience, SDK/module devs rarely care about this and just want to get their idea/app implemented as cleanly and quickly as possible. Most ppl don't care or even know about ocaps.

Likely point 1. above (about a stable SDK) is a bigger reason. As a user of the SDK when I was trying to develop modules, this felt problematic. I would say that lack of proper encapsulation added to the feeling that I was building on quicksand because it was obvious where there needed to be better separation of concerns. That said, I actually think it would be useful to change this ADR to have a broader framing around "protobuf-based inter-module communication" which addresses a stable interface and module accounts in addition to ocaps.

Reply:
@fedekunze I'll think more about the approach in https://github.com/cosmos/cosmos-sdk/issues/5931 and write up a little to compare and contrast.

Reply:
So I think the main advantages of this approach over `x/capability` are how it's integrated with other pieces of the SDK, specifically:
* protobuf so that:
  * code generation of interfaces can be leveraged for a better dev UX
  * module interfaces are versioned and checked for breakage using [buf](https://docs.buf.build/breaking-overview)
* sub-module accounts as per ADR 028
* the general `Msg` passing paradigm and the way signers are specified by `GetSigners`

Also, it's a complete replacement for keepers and could be applied to _all_ inter-module communication whereas the `x/capability` approach in #5931 would need to be applied method by method

Reply:
Updated the ADR with this comparison